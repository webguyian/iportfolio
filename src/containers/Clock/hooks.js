import { useEffect, useRef, useState } from 'react';
import { useHistory } from 'react-router-dom';

import { initialValues } from './constants';
import { getExpiration, getExpirationValues, getSeconds } from './helpers';

export const useInitialRoute = (match, path) => {
  const history = useHistory();

  useEffect(() => {
    if (match.isExact) {
      history.push(path);
    }
  }, []);

  return history;
};

export const useStorage = (key, currentValue, deleteKey, keepUpdating) => {
  const getValue = () => {
    try {
      // Get item from local storage with key
      const item = window.localStorage.getItem(key);

      // Parse stored JSON or if none return undefined
      return item ? JSON.parse(item) : undefined;
    } catch (error) {
      // If error return undefined
      return undefined;
    }
  };

  const value = getValue();
  const localRef = useRef(value);
  const [storedValue, setStoredValue] = useState(value);

  const setValue = nextValue => {
    try {
      // Save to state and local storage
      setStoredValue(nextValue);

      if (deleteKey && !nextValue[deleteKey]) {
        // Remove item from local storage
        window.localStorage.removeItem(key);
      } else {
        // Update local storage and include timestamp
        nextValue.timestamp = Date.now();
        window.localStorage.setItem(key, JSON.stringify(nextValue));
      }
    } catch (error) {
      // Drop error
    }
  };

  useEffect(() => {
    // Update local value
    localRef.current = currentValue || storedValue;

    if (keepUpdating) {
      // Set update value in local storage
      setValue(localRef.current);
    }
  }, [currentValue, storedValue]);

  useEffect(() => {
    return () => {
      if (key && localRef.current) {
        setValue(localRef.current);
      }
    };
  }, []);

  return localRef.current;
};

export const useLaps = (initialLaps, timer, timerRunning) => {
  const [laps, setLaps] = useState(initialLaps);
  const hasLaps = laps.some(lap => lap > 0);
  const storage = useStorage(
    'stopwatch',
    {
      laps,
      timer,
      timerRunning
    },
    'timer'
  );

  const updateLaps = () => {
    if (timerRunning) {
      const hasEmptyLaps = laps.some(lap => lap === 0);

      if (hasEmptyLaps) {
        setLaps(updatedLaps => [
          0,
          ...updatedLaps.slice(0, updatedLaps.length - 1)
        ]);
      } else {
        setLaps(updatedLaps => [0, ...updatedLaps]);
      }
    } else {
      setLaps(initialLaps);
    }
  };

  useEffect(() => {
    if (storage && storage.laps) {
      setLaps(storage.laps);
    }
  }, []);

  useEffect(() => {
    if (!timerRunning || !timer) {
      return;
    }

    const lapLength = laps.length;

    if (!hasLaps && timer > 10) {
      setLaps(updatedLaps => [timer, ...updatedLaps]);
    }

    if (lapLength > 5) {
      setLaps(updatedLaps => {
        const truncatedLaps = updatedLaps.slice(1, lapLength);
        const offset = truncatedLaps.reduce((a, b) => a + b, 0);

        return [timer - offset, ...truncatedLaps];
      });
    }
  }, [timer, timerRunning, hasLaps]);

  return [laps, updateLaps, hasLaps];
};

export const useStopwatch = (startTime = 0, start = false) => {
  const [timerStarted, startTimer] = useState(start);
  const [interval, setLocalInterval] = useState(null);
  const [offset, setOffset] = useState(0);
  const timer = useRef(startTime);
  const storage = useStorage('stopwatch');
  const toggleTimer = () => startTimer(!timerStarted);
  const resetTimer = () => {
    timer.current = 0;
    return timer.current;
  };
  const getUpdatedTime = cache => {
    const now = Date.now();
    const diff = now - cache.timestamp;

    return cache.timer + diff;
  };

  useEffect(() => {
    if (storage && storage.timer) {
      // Get updated time from local storage
      if (storage.timerRunning) {
        timer.current = getUpdatedTime(storage);
        toggleTimer();
      } else {
        timer.current = storage.timer;
      }
    }
  }, []);

  useEffect(() => {
    let localOffset = offset;

    if (timerStarted && !interval) {
      localOffset = Date.now();

      const getDelta = () => {
        const now = Date.now();
        const delta = now - localOffset;

        localOffset = now;

        setOffset(now);
        return delta;
      };

      const localInterval = window.setInterval(() => {
        timer.current += getDelta();
      }, 1);

      setLocalInterval(localInterval);
    }

    if (!timerStarted && interval) {
      window.clearInterval(interval);
      setLocalInterval(null);
    }
  }, [timerStarted, timer, interval]);

  return [timer.current, timerStarted, toggleTimer, resetTimer];
};

export const useDuration = () => {
  const [values, setValues] = useState(null);
  const storage = useStorage('duration', values, null, true);
  const setDuration = (key, value) => {
    setValues({
      ...values,
      [key]: value
    });
  };
  const cachedValues = storage || values;
  const duration = cachedValues && getSeconds(cachedValues);

  useEffect(() => {
    if (storage) {
      // Update state with values from local storage
      setValues(storage);
    } else {
      // Set default initial values
      setValues(initialValues);
    }
  }, []);

  return [duration, setDuration, cachedValues, setValues];
};

export const useCountdown = (h = 0, m = 0, s = 0) => {
  const [seconds, setSeconds] = useState(s);
  const [minutes, setMinutes] = useState(m);
  const [hours, setHours] = useState(h);
  const allSeconds = getSeconds({ hours, minutes, seconds });

  const subtractHour = () => {
    setHours(prevHours => {
      if (prevHours === 0) {
        return 23;
      } else if (prevHours > 0) {
        return prevHours - 1;
      }
      return 0;
    });
  };

  const subtractMinute = () => {
    setMinutes(prevMinutes => {
      if (prevMinutes === 0) {
        subtractHour();
        return 59;
      } else if (prevMinutes > 0) {
        return prevMinutes - 1;
      }
      return 0;
    });
  };

  const subtractSecond = () => {
    setSeconds(prevSeconds => {
      if (prevSeconds === 0) {
        subtractMinute();
        return 59;
      } else if (prevSeconds > 0) {
        return prevSeconds - 1;
      }
      return 0;
    });
  };

  const setCountdown = values => {
    if (!values) {
      setHours(0);
      setMinutes(0);
      setSeconds(0);
      return;
    }

    setHours(values.hours);
    setMinutes(values.minutes);
    setSeconds(values.seconds);
  };

  return [
    {
      allSeconds,
      hours,
      minutes,
      seconds
    },
    setCountdown,
    subtractSecond
  ];
};

export const useTimer = duration => {
  const [started, setStarted] = useState(false);
  const [running, setRunning] = useState(false);
  const [countdown, setCountdown, startCountdown] = useCountdown();
  const intervalRef = useRef();
  const storage = useStorage(
    'timer',
    {
      ...countdown,
      duration,
      running,
      started
    },
    'allSeconds'
  );

  const pause = () => {
    if (intervalRef.current) {
      setRunning(() => false);
      clearInterval(intervalRef.current);
      intervalRef.current = undefined;
    }
  };

  const reset = () => {
    pause();
    setCountdown(null);
    setStarted(false);
  };

  const calculateExpiration = expiration => {
    const distance = expiration - Date.now();
    const expirationValues = getExpirationValues(distance);

    if (countdown.allSeconds < 0) {
      reset();
    } else {
      setCountdown(expirationValues);
    }

    return expirationValues;
  };

  const start = expiration => {
    if (expiration && !intervalRef.current) {
      calculateExpiration(expiration);
    }
  };

  const resume = () => {
    if (!intervalRef.current) {
      setStarted(() => true);
      setRunning(() => true);
      intervalRef.current = setInterval(startCountdown, 1000);
    }
  };

  const restart = () => {
    const expiration = getExpiration(duration);

    reset();
    start(expiration);
  };

  const toggle = () => {
    if (!intervalRef.current) {
      resume();
    } else {
      pause();
    }
  };

  useEffect(() => {
    const expiration = getExpiration(duration, storage);

    start(expiration);

    if (storage) {
      if (storage.running) {
        resume();
      } else if (storage.started) {
        setStarted(true);
      }

      if (duration !== storage.duration) {
        // Remove timer from local storage
        window.localStorage.removeItem('timer');
      }
    }

    return pause;
  }, [duration]);

  return {
    ...countdown,
    duration,
    start,
    started,
    pause,
    reset,
    resume,
    restart,
    running,
    toggle
  };
};

export const useRadialOffset = (timer, radius) => {
  const circumference = (2 * radius * 22) / 7;
  const { allSeconds, duration } = timer;
  const [initialSeconds] = useState(allSeconds);
  const [offset, setOffset] = useState(0);

  useEffect(() => {
    if (!offset) {
      const initialOffset = (allSeconds / duration) * circumference;
      const updatedOffset = circumference - Math.round(initialOffset);

      if (updatedOffset === circumference || updatedOffset < 0) {
        // Exit early
        return;
      }

      setOffset(updatedOffset);
    }
  }, [allSeconds, duration, offset]);

  return [offset, initialSeconds];
};
