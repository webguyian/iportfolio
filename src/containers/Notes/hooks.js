import { useEffect, useRef, useState } from 'react';
import { useSwipeable } from 'react-swipeable';

export const useRefFocus = (condition = true, initial = null) => {
  const ref = useRef(initial);

  useEffect(() => {
    const element = ref && ref.current;

    if (condition && element) {
      // Add focus to element
      element.focus();
    }
  }, [condition, ref]);

  return ref;
};

export const useSwipeOffset = (initial = 0) => {
  const [offset, setOffset] = useState(initial);
  const maxOffset = -30;

  const handleSwipe = eventInfo => {
    if (eventInfo.dir !== 'Down') {
      // Exit early if not down swipe
      return false;
    }

    setOffset(eventInfo.deltaY);
  };

  const handleSwipeUp = () => {
    if (!offset) {
      // Exit early with no offset
      return false;
    }

    setOffset(0);
  };

  const handlers = useSwipeable({
    onSwipedUp: handleSwipeUp,
    onSwiping: handleSwipe,
    trackMouse: true
  });

  return [handlers, offset < maxOffset, setOffset];
};
