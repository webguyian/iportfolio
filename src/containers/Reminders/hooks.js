import { useEffect, useState, useRef } from 'react';
import { useSwipeable } from 'react-swipeable';

export const useLocalStorage = (key, initialValue) => {
  const getValue = () => {
    try {
      // Get item from local storage with key
      const item = window.localStorage.getItem(key);

      // Parse stored json or if none return initialValue
      return item ? JSON.parse(item) : initialValue;
    } catch (error) {
      // If error return initialValue
      return initialValue;
    }
  };
  const [storedValue, setStoredValue] = useState(getValue);
  const setValue = value => {
    try {
      // Allow value to be a function so we have same API as useState
      const valueToStore =
        value instanceof Function ? value(storedValue) : value;

      // Save state
      setStoredValue(valueToStore);
      // Save to local storage
      window.localStorage.setItem(key, JSON.stringify(valueToStore));
    } catch (error) {
      // Drop error
    }
  };

  return [storedValue, setValue, getValue];
};

export const useRefFocus = (node = null) => {
  const ref = useRef(node);

  useEffect(() => {
    const element = ref && ref.current;

    if (element) {
      // Add focus to element
      element.focus();
    }
  }, [ref]);

  return ref;
};

export const useRefControlledFocus = (focused, node = null) => {
  const ref = useRef(node);

  useEffect(() => {
    const element = ref && ref.current;

    if (element && focused) {
      element.focus();
    }
  }, [focused]);

  return ref;
};

export const useReminders = (reminders, setReminders) => {
  const remindersRef = useRef(reminders);

  useEffect(() => {
    // Update reminders in ref
    remindersRef.current = reminders;
  });

  useEffect(() => {
    return () => {
      const { current } = remindersRef;
      const updatedReminders = current.filter(reminder =>
        reminder.value.trim()
      );

      // Trigger callback without empty reminders
      setReminders(updatedReminders);
    };
  }, [remindersRef]);

  return remindersRef;
};

export const useSwipeOffset = (callback, initial = 0) => {
  const [offset, setOffset] = useState(initial);

  const handleSwipe = eventInfo => {
    if (eventInfo.dir !== 'Left') {
      // Exit early if not left swipe
      return false;
    }

    setOffset(eventInfo.deltaX);
  };

  const handleSwipeLeft = eventInfo => {
    const { deltaX } = eventInfo;

    if (deltaX > 80) {
      callback(eventInfo);
    }
  };

  const handleSwipeRight = () => {
    if (!offset) {
      // Exit early with no offset
      return false;
    }

    setOffset(0);
  };

  const handlers = useSwipeable({
    onSwipedLeft: handleSwipeLeft,
    onSwipedRight: handleSwipeRight,
    onSwiping: handleSwipe,
    trackMouse: true
  });

  return [handlers, offset, setOffset];
};
