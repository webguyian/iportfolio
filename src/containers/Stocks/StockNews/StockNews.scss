.stock-news {
  position: absolute;
  bottom: 0;
  left: 0;
  min-height: 100px;
  margin: 0 $gutter-xs;
  transform: translateY(86%);
  transition: transform 0.4s ease;
  border-radius: $border-radius-m;
  background-color: $black-dark;

  .stocks-app--with-search & {
    transform: translateY(110%);
  }

  &--visible {
    transform: translateY(0);
  }

  &-drag-handle {
    display: block;
    width: 40px;
    margin: $gutter-s auto;
    padding: 3px;
    border-radius: $border-radius-m;
    background-color: $gray-medium;
  }

  &-header {
    padding: 0 $gutter $gutter;
    border-bottom: 1px solid $gray-medium;
  }

  &-title {
    font-size: $font-size-l;
  }

  &-author {
    margin: 0;
    font-size: $font-size-m;
  }

  &-list {
    @include unstyled-list;
    max-height: 620px;
    overflow-y: auto;
  }

  &-source {
    font-size: $font-size-s;
  }

  &-headline {
    margin-bottom: $gutter-s;
    font-size: $font-size-s;
  }

  &-timestamp {
    display: block;
    margin-top: auto;
    color: $gray;
    font-size: $font-size-xs;
  }
}
