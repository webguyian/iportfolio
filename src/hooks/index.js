import { useEffect, useState } from 'react';

export const useBreakpoint = breakpoint => {
  const [matchesBreakpoint, setBreakpoint] = useState(false);
  const mediaQuery = `(max-width: ${breakpoint}px)`;

  const handleResize = () => {
    const matches = window.matchMedia(mediaQuery).matches;

    setBreakpoint(matches);
  };

  useEffect(() => {
    window.addEventListener('resize', handleResize);

    return () => {
      window.removeEventListener('resize', handleResize);
    };
  }, [breakpoint]);

  return matchesBreakpoint;
};
