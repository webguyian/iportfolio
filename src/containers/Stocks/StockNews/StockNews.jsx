import React from 'react';
import { string, number, bool } from 'prop-types';
import cx from 'classnames';

import { formatTimestamp } from 'modules/stocks/helpers';
import { useStockNews } from 'modules/stocks/hooks';

import Button from 'components/Button/Button';
import Link from 'components/Link/Link';
import MediaBlock from 'components/MediaBlock/MediaBlock';
import Text from 'components/Text/Text';

const StocksNews = props => {
  const { category, displayCount, draggable } = props;
  const [news, visible, handleSlideUp, handlers] = useStockNews(category);
  const baseClass = 'stock-news';
  const visibleClass = visible && `${baseClass}--visible`;
  const accessibleLabel = visible ? 'drag down' : 'drag up';
  const count = displayCount || (news && news.length);
  const stories = news && news.slice(0, count);

  if (!news || !news.length) {
    return null;
  }

  return (
    <aside className={cx(baseClass, visibleClass)} {...handlers}>
      {draggable && (
        <Button
          className={`${baseClass}-drag-handle`}
          onClick={handleSlideUp}
          aria-labelledby={baseClass}
        >
          <Text type="accessible">{accessibleLabel}</Text>
        </Button>
      )}
      <header id={baseClass} className={`${baseClass}-header`}>
        {!category && (
          <Text className={`${baseClass}-title`} element="h1" type="display">
            Business News
          </Text>
        )}
        <Text className={`${baseClass}-author`} element="h2">
          from Finnhub
        </Text>
      </header>
      {news && (
        <ul className={`${baseClass}-list`}>
          {stories.map(story => (
            <li key={story.id} className={`${baseClass}-list-item`}>
              <Link to={story.url} external>
                <MediaBlock align="right" src={story.image}>
                  <Text className={`${baseClass}-source`} type="display">
                    {story.source}
                  </Text>
                  <Text className={`${baseClass}-headline`}>
                    {story.headline}
                  </Text>
                  <Text className={`${baseClass}-timestamp`}>
                    {formatTimestamp(story.datetime)}
                  </Text>
                </MediaBlock>
              </Link>
            </li>
          ))}
        </ul>
      )}
    </aside>
  );
};

StocksNews.propTypes = {
  category: string,
  displayCount: number,
  draggable: bool
};

StocksNews.defaultProps = {
  draggable: false
};

export default StocksNews;
