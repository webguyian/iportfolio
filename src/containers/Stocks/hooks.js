import { useEffect, useState } from 'react';

import { useStorageCache } from 'hooks';
import { API_EXCHANGE, API_TOKEN } from 'containers/Stocks/constants';
import {
  getCandlestickEndpoint,
  getCandlestickEndpoints,
  getStockEndpoint,
  getStockNewsEndpoint,
  isExpired,
  isExpiredNews,
  updateChartData,
  rehydrateChartData
} from 'containers/Stocks/helpers';

export const useFetch = endpoint => {
  const [data, setData] = useState(null);
  const fetchData = async () => {
    const response = await fetch(endpoint);
    const result = await response.json();

    setData(result);
  };

  useEffect(() => {
    if (!endpoint) {
      return;
    }

    fetchData();
  }, [endpoint]);

  return data;
};

export const useFetchAll = urls => {
  const [responses, setResponses] = useState([]);

  useEffect(() => {
    const fetchData = async () => {
      const results = await Promise.all(
        urls.map(async url => {
          const response = await fetch(url);

          return response.json();
        })
      );

      setResponses(results);
    };

    if (urls.length) {
      fetchData();
    }
  }, [urls]);

  return responses;
};

export const useSearch = () => {
  const [value, setValue] = useState('');
  const [searchTerm, setSearchTerm] = useState('');
  const [searchFocus, setSearchFocus] = useState(false);
  const hasSearch = Boolean(searchFocus || searchTerm.length);

  const handleCancel = () => {
    setValue('');
    setSearchTerm('');
    setSearchFocus(false);
  };

  const handleChange = event => {
    setValue(event.target.value);
  };

  const handleFocus = () => {
    setSearchFocus(focus => !focus);
  };

  const searchHandlers = {
    onBlur: handleFocus,
    onCancel: handleCancel,
    onChange: handleChange,
    onFocus: handleFocus,
    value
  };

  useEffect(() => {
    if (value) {
      setSearchTerm(value.toUpperCase().trim());
    } else {
      setSearchTerm('');
    }
  }, [value]);

  return [searchTerm, searchHandlers, hasSearch];
};

export const useStockSearch = (allStocks, searchTerm) => {
  const [filteredStocks, setFilteredStocks] = useState(allStocks);

  useEffect(() => {
    if (searchTerm.length > 1) {
      const filtered = allStocks.filter(stock => {
        return (
          stock.symbol.indexOf(searchTerm) === 0 ||
          stock.description.indexOf(searchTerm) === 0
        );
      });

      setFilteredStocks(filtered);
    } else if (!searchTerm) {
      setFilteredStocks(allStocks);
    }
  }, [searchTerm]);

  return filteredStocks;
};

export const useStocks = initialStocks => {
  const [endpoints, setEndpoints] = useState([]);
  const [stocks, setStocks] = useState([]);
  const [allStocks, setAllStocks] = useState([]);
  const [activeStocks, setActiveStocks] = useState([]);
  const exchangeResponse = useFetch(API_EXCHANGE);
  const response = useFetchAll(endpoints);
  const cache = useStorageCache(
    'stocks',
    { stocks },
    stockData => !stockData.stocks.length
  );

  useEffect(() => {
    const cachedStocks = cache && cache.stocks;

    if (cachedStocks && cachedStocks.length) {
      setStocks(cachedStocks);
      setActiveStocks(cachedStocks);
    } else {
      setStocks(initialStocks);
      setActiveStocks(initialStocks);
    }
  }, []);

  useEffect(() => {
    if (!exchangeResponse) {
      // Exit early if no response
      return;
    }

    const symbols = activeStocks.map(stock => stock.symbol);
    const currentStocks = exchangeResponse.filter(stock =>
      symbols.includes(stock.symbol)
    );

    setEndpoints(currentStocks.map(getStockEndpoint));
    setAllStocks(exchangeResponse);
  }, [exchangeResponse]);

  useEffect(() => {
    if (activeStocks.length) {
      setEndpoints(activeStocks.map(getStockEndpoint));
    }
  }, [activeStocks]);

  useEffect(() => {
    if (!response.length || !activeStocks.length) {
      return;
    }

    const updatedStocks = response.map((result, index) => {
      const stock = activeStocks[index];

      return {
        ...result,
        ...stock,
        price: result.c,
        previousPrice: result.pc
      };
    });

    setStocks(updatedStocks);
  }, [response, activeStocks]);

  return [stocks, setStocks, allStocks];
};

export const useStockDetail = () => {
  const [endpoint, setEndpoint] = useState('');
  const [activeStock, setActiveStock] = useState(null);
  const [stock, setStockDetail] = useState(activeStock);
  const response = useFetch(endpoint);
  const getStock = symbol =>
    `https://finnhub.io/api/v1/quote?symbol=${symbol}&token=${API_TOKEN}`;

  useEffect(() => {
    setStockDetail(activeStock);

    if (!activeStock) {
      return;
    }

    if (!activeStock.price) {
      setEndpoint(getStock(activeStock.symbol));
    }
  }, [activeStock]);

  useEffect(() => {
    if (response) {
      setStockDetail({
        ...activeStock,
        ...response,
        price: response.c,
        previousPrice: response.pc
      });
      setEndpoint('');
    }
  }, [response]);

  return [stock, setActiveStock];
};

export const useStockNews = category => {
  const [endpoint, setEndpoint] = useState('');
  const [news, setNews] = useState([]);
  const response = useFetch(endpoint);
  const cache = useStorageCache(
    'stock-news',
    { news: response },
    res => !res.news.length
  );

  useEffect(() => {
    const cachedNews = cache && cache.news;

    if (cachedNews && cachedNews.length) {
      setNews(cachedNews);
      const expired = isExpiredNews(cachedNews[0].datetime);

      if (!expired && !category) {
        // Exit early when cache is not expired and no category
        return;
      }
    }

    setEndpoint(getStockNewsEndpoint(category));
  }, [category]);

  useEffect(() => {
    if (response && response.length) {
      setNews(response);
    }
  }, [response]);

  return news;
};

export const useStockChart = (symbol, range) => {
  const [endpoint, setEndpoint] = useState();
  const [updatedResponse, setResponse] = useState(null);
  const response = useFetch(endpoint);

  useEffect(() => {
    const updatedEndpoint = getCandlestickEndpoint(symbol, range);

    setEndpoint(updatedEndpoint);
  }, [symbol, range]);

  useEffect(() => {
    if (response && response.s === 'ok') {
      const data = updateChartData(response, range);

      setResponse(data);
    }
  }, [response]);

  return updatedResponse;
};

export const useStockTicker = (stocks, range = '1D') => {
  const [endpoints, setEndpoints] = useState([]);
  const [updatedResponse, setResponse] = useState([]);
  const response = useFetchAll(endpoints);
  const cache = useStorageCache(
    'stock-charts',
    { charts: updatedResponse },
    res => !res.charts.length
  );

  useEffect(() => {
    const cachedCharts = cache && cache.charts;

    if (cachedCharts && cachedCharts.length) {
      const updated = cachedCharts.map(rehydrateChartData);
      const [head] = updated;
      const expired = isExpired(head.chartData.endDate);

      setResponse(updated);

      if (!expired) {
        // Exit early when cache is not expired
        return;
      }
    }

    // Trigger fetch for candlestick data
    setEndpoints(getCandlestickEndpoints(stocks, range));
  }, [stocks.length]);

  useEffect(() => {
    if (response.length) {
      const updated = response.map((res, index) => {
        if (res.s !== 'ok') {
          return null;
        }

        const stock = stocks[index];
        const chartData = updateChartData(res, range);

        return {
          ...stock,
          chartData
        };
      });

      if (updated[0] !== null) {
        setResponse(updated);
      }
    }
  }, [response]);

  return updatedResponse;
};
