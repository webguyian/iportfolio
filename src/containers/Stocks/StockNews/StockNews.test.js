import React from 'react';
import { act, create } from 'react-test-renderer';

import * as hooks from 'modules/stocks/hooks';
import StockNews from './StockNews';

describe('<StockNews />', () => {
  const news = [
    {
      category: 'top news',
      datetime: 1581389291,
      headline:
        "Kobe Bryant's wife Vanessa wishes 'this nightmare would be over'",
      id: 958565,
      image:
        'https://s4.reutersmedia.net/resources/r/?m=02&d=20200211&t=2&i=1488599576&w=1200&r=LYNXMPEG1A07U',
      related: '',
      source: 'Reuters',
      summary:
        'Vanessa Bryant said in an Instagram post on Monday that she was both grieving and angry over the loss of her husband, NBA star Kobe Bryant, and 13-year-old Gianna in a helicopter crash last month.',
      url: 'https://www.reuters.com/article/us-people-bryant/kobe-bryants-wife-vanessa-wishes-this-nightmare-would-be-over-idUSKBN20507K'
    },
    {
      category: 'top news',
      datetime: 1581389048,
      headline:
        'Coronavirus case confirmed in California, takes U.S. total to 13',
      id: 958566,
      image: 'https://s4.reutersmedia.net/resources_v2/images/rcom-default.png',
      related: '',
      source: 'Reuters',
      summary:
        'The 13th case of coronavirus in the United States was detected in California in an individual under federal quarantine after returning from Wuhan, China, the U.S. Centers for Disease Control and Prevention (CDC) said on Monday.',
      url: 'https://www.reuters.com/article/us-china-health-usa-sandiego/coronavirus-case-confirmed-in-california-takes-u-s-total-to-13-idUSKBN20505I'
    }
  ];
  const values = [news, false, jest.fn(), {}];

  hooks.useStockNews = jest.fn(() => values);

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('renders correctly without news', () => {
    const component = create(<StockNews />);

    expect(component).toMatchSnapshot();
  });

  it('renders correctly with news', () => {
    hooks.useStockNews.mockReturnValue(values);
    const component = create(<StockNews draggable />);

    expect(component).toMatchSnapshot();
  });

  it('handles slide up', () => {
    hooks.useStockNews.mockReturnValue([news, true, jest.fn(), {}]);
    const component = create(<StockNews draggable />);
    const button = component.root.findByProps({
      className: 'stock-news-drag-handle'
    });

    act(() => {
      button.props.onClick();
    });

    expect(component).toMatchSnapshot();
  });
});
