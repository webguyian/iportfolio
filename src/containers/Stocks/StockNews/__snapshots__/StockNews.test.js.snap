// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<StockNews /> handles slide up 1`] = `
<aside
  className="stock-news stock-news--visible"
>
  <button
    aria-labelledby="stock-news"
    className="ui-btn stock-news-drag-handle"
    onClick={[Function]}
    type="button"
  >
    <span
      className="ui-text ui-text--accessible"
    >
      drag down
    </span>
  </button>
  <header
    className="stock-news-header"
    id="stock-news"
  >
    <h1
      className="ui-text ui-text--display stock-news-title"
    >
      Business News
    </h1>
    <h2
      className="ui-text ui-text--body stock-news-author"
    >
      from Finnhub
    </h2>
  </header>
  <ul
    className="stock-news-list"
  >
    <li
      className="stock-news-list-item"
    >
      <Link
        external={true}
        to="https://www.reuters.com/article/us-people-bryant/kobe-bryants-wife-vanessa-wishes-this-nightmare-would-be-over-idUSKBN20507K"
      >
        <div
          className="ui-media-block ui-media-block--right"
        >
          <div
            className="ui-media-block-content"
          >
            <span
              className="ui-text ui-text--display stock-news-source"
            >
              Reuters
            </span>
            <span
              className="ui-text ui-text--body stock-news-headline"
            >
              Kobe Bryant's wife Vanessa wishes 'this nightmare would be over'
            </span>
            <span
              className="ui-text ui-text--body stock-news-timestamp"
            >
              less than a minute ago
            </span>
          </div>
          <img
            className="ui-media-block-image"
            src="https://s4.reutersmedia.net/resources/r/?m=02&d=20200211&t=2&i=1488599576&w=1200&r=LYNXMPEG1A07U"
          />
        </div>
      </Link>
    </li>
    <li
      className="stock-news-list-item"
    >
      <Link
        external={true}
        to="https://www.reuters.com/article/us-china-health-usa-sandiego/coronavirus-case-confirmed-in-california-takes-u-s-total-to-13-idUSKBN20505I"
      >
        <div
          className="ui-media-block ui-media-block--right"
        >
          <div
            className="ui-media-block-content"
          >
            <span
              className="ui-text ui-text--display stock-news-source"
            >
              Reuters
            </span>
            <span
              className="ui-text ui-text--body stock-news-headline"
            >
              Coronavirus case confirmed in California, takes U.S. total to 13
            </span>
            <span
              className="ui-text ui-text--body stock-news-timestamp"
            >
              less than a minute ago
            </span>
          </div>
          <img
            className="ui-media-block-image"
            src="https://s4.reutersmedia.net/resources_v2/images/rcom-default.png"
          />
        </div>
      </Link>
    </li>
  </ul>
</aside>
`;

exports[`<StockNews /> renders correctly with news 1`] = `
<aside
  className="stock-news"
>
  <button
    aria-labelledby="stock-news"
    className="ui-btn stock-news-drag-handle"
    onClick={[Function]}
    type="button"
  >
    <span
      className="ui-text ui-text--accessible"
    >
      drag up
    </span>
  </button>
  <header
    className="stock-news-header"
    id="stock-news"
  >
    <h1
      className="ui-text ui-text--display stock-news-title"
    >
      Business News
    </h1>
    <h2
      className="ui-text ui-text--body stock-news-author"
    >
      from Finnhub
    </h2>
  </header>
  <ul
    className="stock-news-list"
  >
    <li
      className="stock-news-list-item"
    >
      <Link
        external={true}
        to="https://www.reuters.com/article/us-people-bryant/kobe-bryants-wife-vanessa-wishes-this-nightmare-would-be-over-idUSKBN20507K"
      >
        <div
          className="ui-media-block ui-media-block--right"
        >
          <div
            className="ui-media-block-content"
          >
            <span
              className="ui-text ui-text--display stock-news-source"
            >
              Reuters
            </span>
            <span
              className="ui-text ui-text--body stock-news-headline"
            >
              Kobe Bryant's wife Vanessa wishes 'this nightmare would be over'
            </span>
            <span
              className="ui-text ui-text--body stock-news-timestamp"
            >
              less than a minute ago
            </span>
          </div>
          <img
            className="ui-media-block-image"
            src="https://s4.reutersmedia.net/resources/r/?m=02&d=20200211&t=2&i=1488599576&w=1200&r=LYNXMPEG1A07U"
          />
        </div>
      </Link>
    </li>
    <li
      className="stock-news-list-item"
    >
      <Link
        external={true}
        to="https://www.reuters.com/article/us-china-health-usa-sandiego/coronavirus-case-confirmed-in-california-takes-u-s-total-to-13-idUSKBN20505I"
      >
        <div
          className="ui-media-block ui-media-block--right"
        >
          <div
            className="ui-media-block-content"
          >
            <span
              className="ui-text ui-text--display stock-news-source"
            >
              Reuters
            </span>
            <span
              className="ui-text ui-text--body stock-news-headline"
            >
              Coronavirus case confirmed in California, takes U.S. total to 13
            </span>
            <span
              className="ui-text ui-text--body stock-news-timestamp"
            >
              less than a minute ago
            </span>
          </div>
          <img
            className="ui-media-block-image"
            src="https://s4.reutersmedia.net/resources_v2/images/rcom-default.png"
          />
        </div>
      </Link>
    </li>
  </ul>
</aside>
`;

exports[`<StockNews /> renders correctly without news 1`] = `null`;
